File: src/engine/executors/script-executor.js
"use strict";

const fs = require("fs");
const path = require("path");
const { spawn } = require("child_process");

const ALLOWED_INTERPRETERS = {
  bash: "bash",
  python3: "python3",
  node: "node",
};
const EXTENSION_INTERPRETER = {
  ".sh": "bash",
  ".py": "python3",
  ".js": "node",
};

class ScriptSecurityError extends Error {}

function hasDotDot(rawPath) {
  return rawPath.split(/[\\/]+/).includes("..");
}

function resolveScriptPath(repoRoot, rawPath) {
  if (path.isAbsolute(rawPath) || hasDotDot(rawPath)) {
    throw new ScriptSecurityError(
      `script path '${rawPath}' must be relative, under scripts/, with no '..' segment`
    );
  }
  const repoRootResolved = fs.realpathSync(repoRoot);
  const scriptsRootResolved = fs.realpathSync(path.join(repoRootResolved, "scripts"));
  const candidate = path.resolve(repoRootResolved, rawPath);
  const rel = path.relative(scriptsRootResolved, candidate);
  if (rel.startsWith("..") || path.isAbsolute(rel)) {
    throw new ScriptSecurityError(`script path '${rawPath}' must be under scripts/`);
  }

  let walked = repoRootResolved;
  for (const part of rawPath.split(/[\\/]+/)) {
    walked = path.join(walked, part);
    if (fs.existsSync(walked) && fs.lstatSync(walked).isSymbolicLink()) {
      throw new ScriptSecurityError(`script path '${rawPath}' contains a symlink - not allowed`);
    }
  }

  let stat;
  try {
    stat = fs.statSync(candidate);
  } catch (err) {
    throw new ScriptSecurityError(`script '${rawPath}' does not exist or is not a regular file`);
  }
  if (!stat.isFile()) {
    throw new ScriptSecurityError(`script '${rawPath}' does not exist or is not a regular file`);
  }
  return fs.realpathSync(candidate);
}

function resolveInterpreter(spec) {
  let interpreter = spec.interpreter;
  if (!interpreter) interpreter = EXTENSION_INTERPRETER[path.extname(spec.path)];
  if (!Object.prototype.hasOwnProperty.call(ALLOWED_INTERPRETERS, interpreter)) {
    throw new ScriptSecurityError(`interpreter '${interpreter}' is not in the allowlist`);
  }
  return interpreter;
}

function runProcess(command, args, timeoutMs) {
  return new Promise((resolve) => {
    const child = spawn(command, args, { shell: false, stdio: ["ignore", "pipe", "pipe"] });
    let settled = false;
    const timer = setTimeout(() => {
      if (!settled) child.kill("SIGTERM");
      settled = true;
      resolve({ timedOut: true });
    }, timeoutMs);
    child.stdout.resume();
    child.stderr.resume();
    child.on("error", (err) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve({ error: err });
    });
    child.on("close", (code) => {
      if (settled) return;
      settled = true;
      clearTimeout(timer);
      resolve({ code });
    });
  });
}

async function execute(jobId, spec, repoRoot) {
  let scriptPath;
  let interpreter;
  try {
    scriptPath = resolveScriptPath(repoRoot, spec.path);
    interpreter = resolveInterpreter(spec);
  } catch (err) {
    if (err instanceof ScriptSecurityError) return { ok: false, detail: `blocked: ${err.message}` };
    throw err;
  }

  const result = await runProcess(
    ALLOWED_INTERPRETERS[interpreter],
    [scriptPath, ...(spec.args || [])],
    spec.timeout_seconds * 1000
  );
  if (result.timedOut) return { ok: false, detail: `timed out after ${spec.timeout_seconds}s` };
  if (result.error) return { ok: false, detail: `launch error: ${result.error.name || "Error"}` };
  return { ok: result.code === 0, detail: `exit code ${result.code}` };
}

module.exports = {
  ALLOWED_INTERPRETERS,
  ScriptSecurityError,
  resolveScriptPath,
  resolveInterpreter,
  execute,
};

File: engine/executors/script_executor.py
"""Script job execution, confined to the repo's scripts/ directory.

Security model:
  - `script.path` must be a relative path with no leading '/' and no '..'
    segment, and must resolve (after following any symlinks) to somewhere
    under scripts/'s real path — rejects both simple traversal and symlink
    escapes.
  - Only a fixed interpreter allowlist (bash/python3/node) is ever invoked,
    always as an argv list with shell=False — never a shell string, so there
    is no inline-shell-command surface at all.
  - stdout/stderr are captured only to detect success/failure length for the
    detail string's exit code; raw output is never returned or stored (a
    script could easily echo a secret from its environment by accident).
"""
from __future__ import annotations

import os
import subprocess
from pathlib import Path, PurePosixPath

from ..models import ScriptSpec

ALLOWED_INTERPRETERS = {"bash": "bash", "python3": "python3", "node": "node"}

_EXTENSION_INTERPRETER = {".sh": "bash", ".py": "python3", ".js": "node"}


class ScriptSecurityError(Exception):
    pass


def resolve_script_path(repo_root: Path, raw_path: str) -> Path:
    """`raw_path` is repo-relative (e.g. 'scripts/examples/foo.sh') and must
    resolve to somewhere under <repo_root>/scripts/."""
    if raw_path.startswith("/") or ".." in PurePosixPath(raw_path).parts:
        raise ScriptSecurityError(
            f"script path '{raw_path}' must be relative, under scripts/, with no '..' segment"
        )

    repo_root_resolved = repo_root.resolve()
    scripts_root_resolved = (repo_root_resolved / "scripts").resolve()
    candidate = (repo_root_resolved / raw_path).resolve()

    try:
        candidate.relative_to(scripts_root_resolved)
    except ValueError:
        raise ScriptSecurityError(f"script path '{raw_path}' must be under scripts/")

    # Belt-and-suspenders: reject if any path component along the way is a
    # symlink, even though .resolve() above would already have caught an
    # escape via the relative_to() check.
    walked = repo_root_resolved
    for part in Path(raw_path).parts:
        walked = walked / part
        if walked.is_symlink():
            raise ScriptSecurityError(f"script path '{raw_path}' contains a symlink — not allowed")

    if not candidate.is_file():
        raise ScriptSecurityError(f"script '{raw_path}' does not exist or is not a regular file")

    return candidate


def resolve_interpreter(spec: ScriptSpec) -> str:
    interpreter = spec.interpreter
    if interpreter is None:
        ext = Path(spec.path).suffix
        interpreter = _EXTENSION_INTERPRETER.get(ext)
    if interpreter not in ALLOWED_INTERPRETERS:
        raise ScriptSecurityError(f"interpreter '{interpreter}' is not in the allowlist")
    return interpreter


def execute(job_id: str, spec: ScriptSpec, repo_root: Path) -> tuple[bool, str]:
    try:
        script_path = resolve_script_path(repo_root, spec.path)
        interpreter = resolve_interpreter(spec)
    except ScriptSecurityError as e:
        return False, f"blocked: {e}"

    cmd = [ALLOWED_INTERPRETERS[interpreter], str(script_path), *spec.args]
    try:
        proc = subprocess.run(
            cmd,
            shell=False,
            timeout=spec.timeout_seconds,
            capture_output=True,
            text=True,
        )
        return proc.returncode == 0, f"exit code {proc.returncode}"
    except subprocess.TimeoutExpired:
        return False, f"timed out after {spec.timeout_seconds}s"
    except OSError as e:
        return False, f"launch error: {type(e).__name__}"
